### PR TITLE
Open Payment Platform Gateway

### PR DESCRIPTION
**Pull request by borislav-sabev**
HI 

I've implemented Open Payment Platform Gateway - opp upon a request, so this is initial pull  request.

Best regards
Borislav

### CODE REVIEW

File: lib/active_merchant/billing/gateways/opp.rb [hunk 1]
```diff
@@ -0,0 +1,399 @@
...
+    #     :month      => 8,
+    #     :year       => 2009,
+    #     :first_name => 'Bob',
+    #     :last_name  => 'Bobsen'
+    #     :verification_value: '123')
+    #
+    #   # Request: complete example, including address, billing address, shipping address, shopping cart
+    #    complete_request_options = {
+    #      order_id: "your merchant/shop order id", # alternative is to set merchantInvoiceId 
+    #      merchantTransactionId: "your merchant/shop transaction id",
```
Comment (ntalbott) [inline lib/active_merchant/billing/gateways/opp.rb]: All options should be Ruby-style with underscores rather than camel case.

File: lib/active_merchant/billing/gateways/opp.rb [hunk 2]
```diff
@@ -0,0 +1,399 @@
...
+    #              tax: 7.0, shipping: 3.25, discount: 10.0
+    #            }                    
+    #        ],
+    #      },
+    #    }
+    #    
+    #    # Request: minimal example
+    #    minimal_request_options = {
+    #      order_id: "your merchant/shop order id", # alternative is to set merchantInvoiceId 
+    #      merchantTransactionId: "your merchant/shop transaction id",
```
Comment (ntalbott) [inline lib/active_merchant/billing/gateways/opp.rb]: ActiveMerchant doesn't allow non-standard options to be required, so for merging `merchant_transaction_id` must be made optional and/or mapped to a standard option like `order_id`.

File: lib/active_merchant/billing/gateways/opp.rb [hunk 3]
```diff
@@ -0,0 +1,399 @@
...
+
+      def purchase(money, payment, options={})
+        # debit
+        execute_dbpa(options[:riskWorkflow] ? METHOD_TO_PAYMENTTYPE_MAPPING[:pacp]: METHOD_TO_PAYMENTTYPE_MAPPING[__method__], 
+          money, payment, options)
+      end
+
+      def authorize(money, payment, options={})
+        # preauthorization PA
+        execute_dbpa(METHOD_TO_PAYMENTTYPE_MAPPING[__method__], money, payment, options)
```
Comment (ntalbott) [inline lib/active_merchant/billing/gateways/opp.rb]: Not a fan of using `__method__` like this. Just replace with an explicit reference to the method name.

File: lib/active_merchant/billing/gateways/opp.rb [hunk 4]
```diff
@@ -0,0 +1,399 @@
...
+    #        birthDate:  '1965-05-01',
+    #        phone:  '(?!?)555-5555',
+    #        mobile:  '(?!?)234-23423',
+    #        email:  'jane@jones.com',
+    #        companyName:  'JJ Ltd.',
+    #        identificationDocType:  'PASSPORT',
+    #        identificationDocId:  'FakeID2342431234123',
+    #        ip:  101.102.103.104,
+    #      },
+    #      cart: {
```
Comment (ntalbott) [inline lib/active_merchant/billing/gateways/opp.rb]: Cart and line item support are out of scope for ActiveMerchant. Please remove.

File: test/remote/gateways/remote_opp_test.rb [hunk 5]
```diff
@@ -0,0 +1,215 @@
...
+    assert_success refund
+  end
+
+  def test_successful_verify
+    @options[:description] = __method__
+    response = @gateway.verify(@valid_card, @options)
+    assert_success response
+  end
+
+  # def test_purchases_with_registration
```
Comment (ntalbott) [inline test/remote/gateways/remote_opp_test.rb]: Please remove all commented out code.

**Comment by ntalbott** [comment]
@borislav-sabev added comments; needs some work but looks to be headed in the right direction.

**Comment by borislav-sabev** [comment]
Hi ntabott, I've made requested changes see 
https://github.com/borislav-sabev/active_merchant/commit/cc0f02428926716cd2b568812c34d0fff51a1126

Regards
Boris

**Comment by byronkelly** [comment]
Hi,

Just curious when this will be added to the pulled into the branch?

Thank you.

**Comment by timbeiko** [comment]
Shipping in https://github.com/activemerchant/active_merchant/pull/2153 
Closing.
